### PR TITLE
Implement parsing time based on pattern for @ApolloJsonValue

### LINKED ISSUE #49: ApolloJsonValue注解增加部分自定义功能
**现状：**
目前ApolloJsonValue仅支持从portal读取指定配置，没有其他功能支持。在遇到下面场景时显得有些无奈：
假如在配置中有一些日期时间相关的配置，如下：

`{
    "beginTime": "2019-08-21 00:00:00",
    "endTime": "2019-09-05 00:00:00",
    "switch": true
}`

在接收对应的时间按配置时只能用String（如果直接用Date接受，会报错：Failed to parse date ["2019-08-21 00:00:00"]: Invalid time zone indicator ' '），然后在代码中进行格式化。
**建议：**
ApolloJsonValue增加一个 dateFormat ，ApolloJsonValueProcessor在解析式可以用指定的格式去解析

### PR DESCRIPTION
**Pull request by huxleyliau**
## What's the purpose of this PR
Implement parsing time based on pattern for @ApolloJsonValue from #49 

## Which issue(s) this PR fixes:
Fixes #49

## Brief changelog

Make use of Gson's current way of turning a string into a date to make @ApolloJsonValue able to handle time strings according to a given pattern.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Read the [Contributing Guide](https://github.com/apolloconfig/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [x] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [x] Update the [`CHANGES` log](https://github.com/apolloconfig/apollo-java/blob/master/CHANGES.md).

### CODE REVIEW

**Comment by huxleyliau** [comment]
I have read the CLA Document and I hereby sign the CLA

File: apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java [hunk 1]
```diff
@@ -244,9 +250,9 @@ private Object resolvePropertyValue(String beanName, String placeHolder) {
     return propertyValue;
   }
 
-  private Object parseJsonValue(String json, Type targetType) {
+  private Object parseJsonValue(String json, Type targetType, String datePattern) {
     try {
-      return GSON.fromJson(json, targetType);
+      return DATEPATTERN_GSON_MAP.computeIfAbsent(datePattern, (pattern) -> new GsonBuilder().setDateFormat(pattern).create()).fromJson(json, targetType);
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java:255]: The default datePattern value of `Gson` is [null](https://github.com/google/gson/blob/51cce39e0064a755d32f820670770f82519039de/gson/src/main/java/com/google/gson/Gson.java#L267) while the the default datePattern value of `ApolloJsonValue` is empty.
I'm not sure whether they work the same way but it's better to keep the original logic if user has no intention to config the datePattern.

File: apollo-client/src/test/java/com/ctrip/framework/apollo/spring/JavaConfigPlaceholderAutoUpdateTest.java [hunk 2]
```diff
@@ -868,6 +873,7 @@ public void testAutoUpdateWithAllKindsOfDataTypes() throws Exception {
     assertEquals(someDate, bean.getDateProperty());
     assertEquals("astring", bean.getJsonBeanList().get(0).getA());
     assertEquals(10, bean.getJsonBeanList().get(0).getB());
+    assertNotNull(bean.getJsonDateBean().getStartTime());
```
Comment (nobodyiam) [inline apollo-client/src/test/java/com/ctrip/framework/apollo/spring/JavaConfigPlaceholderAutoUpdateTest.java:876]: Please verify the exact start time.

File: apollo-client/src/test/java/com/ctrip/framework/apollo/spring/JavaConfigPlaceholderAutoUpdateTest.java [hunk 3]
```diff
@@ -899,6 +906,7 @@ public void testAutoUpdateWithAllKindsOfDataTypes() throws Exception {
     assertEquals(someNewDate, bean.getDateProperty());
     assertEquals("newString", bean.getJsonBeanList().get(0).getA());
     assertEquals(20, bean.getJsonBeanList().get(0).getB());
+    assertNotNull(bean.getJsonDateBean().getStartTime());
```
Comment (nobodyiam) [inline apollo-client/src/test/java/com/ctrip/framework/apollo/spring/JavaConfigPlaceholderAutoUpdateTest.java:909]: Please verify the exact start time.

File: apollo-client/src/test/java/com/ctrip/framework/apollo/spring/JavaConfigPlaceholderTest.java [hunk 4]
```diff
@@ -385,6 +390,27 @@ public void testApolloJsonValue() {
...
+    when(config.getProperty(eq(DATE_FORMAT_JSON_PROPERTY1), Mockito.nullable(String.class))).thenReturn(dateFormatJson1);
+    when(config.getProperty(eq(DATE_FORMAT_JSON_PROPERTY2), Mockito.nullable(String.class))).thenReturn(dateFormatJson2);
+    when(config.getProperty(eq(DATE_FORMAT_JSON_PROPERTY3), Mockito.nullable(String.class))).thenReturn(dateFormatJson3);
+    mockConfig(ConfigConsts.NAMESPACE_APPLICATION, config);
+
+    AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext(
+        AppConfig12.class);
+
+    TestJsonDatePropertyBean datePropertyBean = context.getBean(TestJsonDatePropertyBean.class);
+    assertNotNull(datePropertyBean.getPattern1());
```
Comment (nobodyiam) [inline apollo-client/src/test/java/com/ctrip/framework/apollo/spring/JavaConfigPlaceholderTest.java:409]: Please verify the exact time.

File: apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java (hunk 1 above)
Comment (huxleyliau) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java:255]: :thumbsup: Yes, it's best to keep the existing logic unchanged.  I'm making the changes now. :smiley:

**Comment by nobodyiam** [comment]
recheck

**Pull request closed without merging**